File: examples/ralf/ralf_subscriber.c
/* Full working RALF subscriber example (C).
 *
 * Goes beyond a single trivial subscription: subscribes to every channel
 * the given role is actually entitled to (protocol reference §7) in one
 * combined SUB, then puts the received EXEC/EOD data to use -- a running
 * per-symbol executed-volume tally that de-duplicates EXEC lines by
 * EXEC_ID (an AUDIT client subscribed to more than one channel receives
 * the *same* trade once per channel, by design, see §6), plus a
 * per-channel sequence-gap check and a clean SIGINT shutdown that prints
 * the final tally before exiting.
 *
 * See docs/user-guide/930-app-ralf-protocol.md for the normative wire
 * contract this client follows.
 */

#define _POSIX_C_SOURCE 200809L

#include "ralf_parser.h"

#include <arpa/inet.h>
#include <errno.h>
#include <netinet/in.h>
#include <signal.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <unistd.h>

#define MAX_CHANNELS 3   /* CLEARING, DROP_COPY, AUDIT */
#define MAX_SYMBOLS 16   /* bounded per-symbol tally table */
#define MAX_EXEC_IDS 256 /* bounded de-dup window for this demo session */

/* SIGINT sets this instead of taking the default terminate action, so the
 * blocking read() in recv_line() can be interrupted cleanly (see the
 * sigaction() call in main(), which deliberately omits SA_RESTART) and
 * the client can print its final tally and close the socket instead of
 * being killed mid-syscall.
 */
static volatile sig_atomic_t g_running = 1;

static void handle_sigint(int signum) {
    (void)signum;
    g_running = 0;
}

/* ------------------------------------------------------------------ */
/* Role -> entitled channel list (protocol reference §7). AUDIT is the
 * only role entitled to more than its own name.
 * ------------------------------------------------------------------ */

static const char *entitled_channels(const char *role) {
    if (strcmp(role, "AUDIT") == 0) {
        return "CLEARING,DROP_COPY,AUDIT";
    }
    return role; /* CLEARING -> "CLEARING", DROP_COPY -> "DROP_COPY" */
}

/* ------------------------------------------------------------------ */
/* Per-channel sequence-gap detection. Unlike CALF, RALF sequences are
 * per-*channel*, not per-(channel,symbol) -- protocol reference §8.
 * ------------------------------------------------------------------ */

typedef struct {
    char channel[RALF_MAX_KEY_LEN];
    long seq;
} channel_seq_t;

static channel_seq_t g_channel_seq[MAX_CHANNELS];
static int g_channel_seq_count = 0;

static void check_sequence(const char *channel, const char *seq_str) {
    if (!channel[0] || !seq_str) {
        return;
    }
    long seq = strtol(seq_str, NULL, 10);

    for (int i = 0; i < g_channel_seq_count; i++) {
        if (strcmp(g_channel_seq[i].channel, channel) == 0) {
            if (seq != g_channel_seq[i].seq + 1) {
                fprintf(stderr, "!! sequence gap on channel %s: expected %ld, got %ld\n", channel,
                        g_channel_seq[i].seq + 1, seq);
            }
            g_channel_seq[i].seq = seq;
            return;
        }
    }
    if (g_channel_seq_count < MAX_CHANNELS) {
        channel_seq_t *entry = &g_channel_seq[g_channel_seq_count++];
        snprintf(entry->channel, sizeof(entry->channel), "%s", channel);
        entry->seq = seq;
    }
}

/* ------------------------------------------------------------------ */
/* Per-symbol executed-volume tally, de-duplicated by EXEC_ID.
 *
 * A single executed trade is delivered once *per subscribed channel*: a
 * role subscribed to more than one channel (only AUDIT can be) sees the
 * same EXEC_ID repeated, once per channel. Counting raw lines would
 * overstate volume by up to 3x for an AUDIT client subscribed to
 * everything, so this tracks seen EXEC_IDs and only tallies each once.
 * ------------------------------------------------------------------ */

typedef struct {
    char symbol[RALF_MAX_VAL_LEN];
    long unique_trades;
    long total_qty;
    int used;
} symbol_tally_t;

static symbol_tally_t g_tallies[MAX_SYMBOLS];

static char g_seen_exec_ids[MAX_EXEC_IDS][RALF_MAX_VAL_LEN];
static int g_seen_exec_id_count = 0;

/* Returns 1 if exec_id was newly recorded, 0 if it was already seen (or
 * the de-dup table is full, in which case it is conservatively treated
 * as a duplicate rather than risk double-counting).
 */
static int record_exec_id(const char *exec_id) {
    for (int i = 0; i < g_seen_exec_id_count; i++) {
        if (strcmp(g_seen_exec_ids[i], exec_id) == 0) {
            return 0;
        }
    }
    if (g_seen_exec_id_count >= MAX_EXEC_IDS) {
        return 0;
    }
    snprintf(g_seen_exec_ids[g_seen_exec_id_count], sizeof(g_seen_exec_ids[0]), "%s", exec_id);
    g_seen_exec_id_count++;
    return 1;
}

static symbol_tally_t *get_tally(const char *symbol) {
    for (int i = 0; i < MAX_SYMBOLS; i++) {
        if (g_tallies[i].used && strcmp(g_tallies[i].symbol, symbol) == 0) {
            return &g_tallies[i];
        }
    }
    for (int i = 0; i < MAX_SYMBOLS; i++) {
        if (!g_tallies[i].used) {
            g_tallies[i].used = 1;
            snprintf(g_tallies[i].symbol, sizeof(g_tallies[i].symbol), "%s", symbol);
            return &g_tallies[i];
        }
    }
    return NULL; /* table full: MAX_SYMBOLS comfortably covers this demo */
}

static void print_summary(void) {
    printf("\nExecution summary: %d unique trade(s)\n", g_seen_exec_id_count);
    for (int i = 0; i < MAX_SYMBOLS; i++) {
        if (g_tallies[i].used) {
            printf("  %-8s %ld unique trade(s), %ld total qty\n", g_tallies[i].symbol, g_tallies[i].unique_trades,
                   g_tallies[i].total_qty);
        }
    }
}

/* ------------------------------------------------------------------ */
/* Line-level socket helpers (identical pattern to calf_subscriber.c). */

static int send_line(int fd, const char *line) {
    size_t len = strlen(line);
    if (write(fd, line, len) != (ssize_t)len) {
        return -1;
    }
    if (write(fd, "\n", 1) != 1) {
        return -1;
    }
    return 0;
}

/* Returns 0 on a complete line, -1 on disconnect/error, -2 if interrupted
 * by SIGINT (see g_running) with no line pending.
 */
static int recv_line(int fd, char *buf, size_t cap) {
    size_t n = 0;
    while (n + 1 < cap) {
        char c;
        ssize_t r = read(fd, &c, 1);
        if (r < 0) {
            if (errno == EINTR) {
                if (!g_running) {
                    return -2;
                }
                continue;
            }
            return -1;
        }
        if (r == 0) {
            return -1; /* peer closed */
        }
        if (c == '\n') {
            break;
        }
        buf[n++] = c;
    }
    buf[n] = '\0';
    return 0;
}

static int connect_gateway(const char *host, int port) {
    int fd = socket(AF_INET, SOCK_STREAM, 0);
    if (fd < 0) {
        return -1;
    }

    struct sockaddr_in addr;
    memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_port = htons((uint16_t)port);

    if (inet_pton(AF_INET, host, &addr.sin_addr) != 1) {
        close(fd);
        return -1;
    }

    if (connect(fd, (struct sockaddr *)&addr, sizeof(addr)) != 0) {
        close(fd);
        return -1;
    }

    return fd;
}

/* ------------------------------------------------------------------ */

static void handle_message(const ralf_message_t *msg) {
    const char *channel = ralf_get_field(msg, "CH");
    const char *symbol = ralf_get_field(msg, "SYM");
    channel = channel ? channel : "";
    symbol = symbol ? symbol : "";
    check_sequence(channel, ralf_get_field(msg, "SEQ"));

    if (strcmp(msg->msg_type, "SNAP") == 0) {
        printf("SNAP  CH=%s SYM=%s SEQ=%s\n", channel, symbol, ralf_get_field(msg, "SEQ"));
    } else if (strcmp(msg->msg_type, "EXEC") == 0) {
        const char *exec_id = ralf_get_field(msg, "EXEC_ID");
        const char *qty_str = ralf_get_field(msg, "QTY");
        const char *px = ralf_get_field(msg, "PX");
        const char *side = ralf_get_field(msg, "SIDE");
        int is_new = exec_id && record_exec_id(exec_id);
        if (is_new) {
            symbol_tally_t *tally = get_tally(symbol);
            if (tally) {
                tally->unique_trades++;
                tally->total_qty += qty_str ? strtol(qty_str, NULL, 10) : 0;
            }
        }
        printf("EXEC  %-8s %6s @ %10s %-4s ch=%s%s\n", symbol, qty_str ? qty_str : "?", px ? px : "?",
               side ? side : "?", channel, is_new ? "" : "  (duplicate copy of an already-counted trade)");
    } else if (strcmp(msg->msg_type, "EOD") == 0) {
        symbol_tally_t *tally = get_tally(symbol);
        const char *trade_count = ralf_get_field(msg, "TRADE_COUNT");
        printf("EOD   %-8s gateway TRADE_COUNT=%s (client tallied %ld unique trade(s), %ld total qty)\n", symbol,
               trade_count ? trade_count : "?", tally ? tally->unique_trades : 0, tally ? tally->total_qty : 0);
    } else if (strcmp(msg->msg_type, "HB") == 0) {
        printf("HB    (gateway heartbeat)\n");
    } else if (strcmp(msg->msg_type, "PONG") == 0) {
        printf("PONG  (liveness reply)\n");
    } else if (strcmp(msg->msg_type, "EXIT") == 0) {
        const char *reason = ralf_get_field(msg, "REASON");
        fprintf(stderr, "EXIT  gateway is closing the session (reason=%s)\n", reason ? reason : "?");
    } else if (strcmp(msg->msg_type, "ERR") == 0) {
        /* RALF's error field is DETAIL, not MSG (CALF's error field name)
         * -- see protocol reference §10.
         */
        const char *code = ralf_get_field(msg, "CODE");
        const char *detail = ralf_get_field(msg, "DETAIL");
        fprintf(stderr, "ERR   %s: %s\n", code ? code : "?", detail ? detail : "");
    } else {
        printf("%s (unhandled, %d field(s))\n", msg->msg_type, msg->field_count);
    }
}

int main(int argc, char **argv) {
    const char *host = "127.0.0.1";
    int port = 5580;
    const char *role = "CLEARING";

    if (argc > 1) {
        host = argv[1];
    }
    if (argc > 2) {
        port = atoi(argv[2]);
    }
    if (argc > 3) {
        role = argv[3];
    }
    if (strcmp(role, "CLEARING") != 0 && strcmp(role, "DROP_COPY") != 0 && strcmp(role, "AUDIT") != 0) {
        fprintf(stderr, "unknown role %s (expected CLEARING, DROP_COPY, or AUDIT)\n", role);
        return 1;
    }

    struct sigaction sa;
    memset(&sa, 0, sizeof(sa));
    sa.sa_handler = handle_sigint;
    sigemptyset(&sa.sa_mask);
    sa.sa_flags = 0; /* no SA_RESTART: interrupt the blocking read() in recv_line() */
    sigaction(SIGINT, &sa, NULL);

    int fd = connect_gateway(host, port);
    if (fd < 0) {
        perror("connect_gateway");
        return 1;
    }

    char hello[256];
    snprintf(hello, sizeof(hello), "HELLO|CLIENT=ext-c-01|PROTO=RALF1|ROLE=%s|LASTSEQ=0", role);
    if (send_line(fd, hello) != 0) {
        perror("send HELLO");
        close(fd);
        return 1;
    }

    char line[4096];
    if (recv_line(fd, line, sizeof(line)) != 0) {
        perror("recv WELCOME");
        close(fd);
        return 1;
    }

    ralf_message_t welcome;
    if (ralf_parse_line(line, &welcome) != 0) {
        fprintf(stderr, "failed to parse WELCOME\n");
        close(fd);
        return 1;
    }
    printf("WELCOME type=%s GW=%s ROLE=%s\n", welcome.msg_type,
           ralf_get_field(&welcome, "GW") ? ralf_get_field(&welcome, "GW") : "?",
           ralf_get_field(&welcome, "ROLE") ? ralf_get_field(&welcome, "ROLE") : "?");

    /* One combined SUB covering every channel the role is entitled to --
     * CH accepts a comma-separated list, so this is one round trip and
     * one SNAP, not one per channel. No ROLE field on SUB: the protocol
     * only defines CH and SYM here (role was already established by
     * HELLO); a gateway would just ignore an extra field, but a "best
     * practice" client should not send fields the wire contract doesn't
     * define.
     */
    char sub_line[128];
    snprintf(sub_line, sizeof(sub_line), "SUB|CH=%s|SYM=*", entitled_channels(role));
    if (send_line(fd, sub_line) != 0) {
        perror("send SUB");
        close(fd);
        return 1;
    }
    printf("Subscribed %s as role=%s\n", sub_line, role);

    for (;;) {
        int rc = recv_line(fd, line, sizeof(line));
        if (rc == -2) {
            printf("\ninterrupted, closing connection\n");
            break;
        }
        if (rc != 0) {
            break; /* disconnected */
        }
        ralf_message_t msg;
        if (ralf_parse_line(line, &msg) != 0) {
            fprintf(stderr, "parse error: %s\n", line);
            continue;
        }
        handle_message(&msg);
    }

    print_summary();
    close(fd);
    return 0;
}

File: docs/examples/ralf/ralf_subscriber.py
#!/usr/bin/env python3
"""Full working RALF subscriber example.

Goes beyond a single trivial subscription: subscribes to every channel a
role is actually entitled to (see the §7 entitlement matrix in the
protocol reference), then puts the received `EXEC`/`EOD` data to use --
a running per-symbol executed-volume tally that correctly de-duplicates
`EXEC` lines by `EXEC_ID` (an `AUDIT` client subscribed to more than one
channel receives the *same* trade once per channel, by design, see §6)
and a per-channel sequence-gap check.

See docs/user-guide/930-app-ralf-protocol.md for the normative wire
contract this client follows.

Uses ralf_parser.py to parse and build RALF lines.
"""

from __future__ import annotations

import argparse
import socket
import sys
from collections.abc import Iterable
from dataclasses import dataclass, field

from ralf_parser import RalfMessage, build_ralf_line, parse_ralf_line

# Which channels each role may subscribe to (protocol reference §7). AUDIT
# is the only role entitled to more than its own name.
_ROLE_CHANNELS: dict[str, tuple[str, ...]] = {
    "CLEARING": ("CLEARING",),
    "DROP_COPY": ("DROP_COPY",),
    "AUDIT": ("CLEARING", "DROP_COPY", "AUDIT"),
}


class LineReader:
    """Buffers raw socket bytes and yields one decoded line at a time.

    TCP is a byte stream, not a message queue: one recv() call may return
    half a line, a whole line, or several lines concatenated together, so
    a real client must buffer and split on '\\n' itself rather than assume
    each recv() lines up with one RALF message.
    """

    def __init__(self, sock: socket.socket) -> None:
        self.sock = sock
        self.buf = bytearray()

    def recv_line(self) -> str:
        while True:
            nl = self.buf.find(b"\n")
            if nl >= 0:
                line = bytes(self.buf[:nl])
                del self.buf[: nl + 1]
                return line.decode("utf-8", errors="replace")

            chunk = self.sock.recv(4096)
            if not chunk:
                raise RuntimeError("gateway closed connection")
            self.buf.extend(chunk)


def send_line(sock: socket.socket, msg_type: str, fields: dict[str, str]) -> None:
    sock.sendall(build_ralf_line(msg_type, fields).encode("utf-8"))


def subscribe(sock: socket.socket, channels: Iterable[str], symbols: str) -> None:
    """One SUB covering every channel in *channels*, all scoped to *symbols*.

    `CH` accepts a comma-separated list (protocol reference §6, `SUB`), so
    an AUDIT client subscribing to all three channels is one round trip
    and one `SNAP`, not three.
    """
    send_line(sock, "SUB", {"CH": ",".join(channels), "SYM": symbols})


class ChannelSequenceTracker:
    """Detects gaps in the per-channel SEQ counters RALF streams use.

    Unlike CALF, RALF sequences are per-*channel*, not per-(channel,
    symbol) -- see §8 of the protocol reference: "SEQ is maintained per
    channel ... across different channels, SEQ values are independent."
    """

    def __init__(self) -> None:
        self._last_seq: dict[str, int] = {}

    def observe(self, channel: str, seq_field: str | None) -> None:
        if not channel or not seq_field:
            return
        try:
            seq = int(seq_field)
        except ValueError:
            return
        previous = self._last_seq.get(channel)
        if previous is not None and seq != previous + 1:
            print(
                f"!! sequence gap on channel {channel}: "
                f"expected {previous + 1}, got {seq}",
                file=sys.stderr,
            )
        self._last_seq[channel] = seq


@dataclass
class SymbolTally:
    unique_trades: int = 0
    total_qty: int = 0


@dataclass
class ExecTracker:
    """Tallies executed volume per symbol from `EXEC` lines.

    A single executed trade is delivered once *per subscribed channel*
    (protocol reference §6, `EXEC`): a role subscribed to more than one
    channel -- only `AUDIT` can be -- sees the same `EXEC_ID` repeated,
    once per channel. Counting raw lines would overstate volume by up to
    3x for an AUDIT client subscribed to everything, so this tracker
    de-duplicates by `EXEC_ID` before tallying and separately reports the
    raw (as-delivered) line count so the distinction is visible rather
    than silently hidden.
    """

    _seen_exec_ids: set[str] = field(default_factory=set)
    _by_symbol: dict[str, SymbolTally] = field(default_factory=dict)
    _raw_lines_by_channel: dict[str, int] = field(default_factory=dict)

    def observe(self, channel: str, fields: dict[str, str]) -> tuple[bool, SymbolTally]:
        """Record one EXEC line; return (is_new_unique_trade, its symbol's tally)."""
        self._raw_lines_by_channel[channel] = (
            self._raw_lines_by_channel.get(channel, 0) + 1
        )
        exec_id = fields.get("EXEC_ID", "")
        symbol = fields.get("SYM", "")
        tally = self._by_symbol.setdefault(symbol, SymbolTally())
        is_new = bool(exec_id) and exec_id not in self._seen_exec_ids
        if is_new:
            self._seen_exec_ids.add(exec_id)
            tally.unique_trades += 1
            try:
                tally.total_qty += int(fields.get("QTY", "0"))
            except ValueError:
                pass
        return is_new, tally

    def tally_for(self, symbol: str) -> SymbolTally:
        return self._by_symbol.get(symbol, SymbolTally())

    def summary(self) -> str:
        raw_total = sum(self._raw_lines_by_channel.values())
        unique_total = len(self._seen_exec_ids)
        by_channel = ", ".join(
            f"{ch}={n}" for ch, n in sorted(self._raw_lines_by_channel.items())
        )
        lines = [
            f"Execution summary: {unique_total} unique trade(s) "
            f"({raw_total} raw EXEC line(s): {by_channel or 'none'})"
        ]
        for symbol, tally in sorted(self._by_symbol.items()):
            lines.append(
                f"  {symbol:<8} {tally.unique_trades} unique trade(s), "
                f"{tally.total_qty} total qty"
            )
        return "\n".join(lines)


def _handle_message(
    msg: RalfMessage, seq_tracker: ChannelSequenceTracker, exec_tracker: ExecTracker
) -> None:
    channel = msg.fields.get("CH", "")
    symbol = msg.fields.get("SYM", "")
    seq_tracker.observe(channel, msg.fields.get("SEQ"))

    if msg.msg_type == "SNAP":
        print(f"SNAP  CH={channel} SYM={symbol} SEQ={msg.fields.get('SEQ', '?')}")
    elif msg.msg_type == "EXEC":
        is_new, _tally = exec_tracker.observe(channel, msg.fields)
        dup_note = "" if is_new else "  (duplicate copy of an already-counted trade)"
        print(
            f"EXEC  {symbol:<8} {msg.fields.get('QTY', '?'):>6} @ "
            f"{msg.fields.get('PX', '?'):>10} {msg.fields.get('SIDE', '?'):<4} "
            f"ch={channel}{dup_note}"
        )
    elif msg.msg_type == "EOD":
        tally = exec_tracker.tally_for(symbol)
        print(
            f"EOD   {symbol:<8} gateway TRADE_COUNT="
            f"{msg.fields.get('TRADE_COUNT', '?')} "
            f"(client tallied {tally.unique_trades} unique trade(s), "
            f"{tally.total_qty} total qty)"
        )
    elif msg.msg_type == "HB":
        print("HB    (gateway heartbeat)")
    elif msg.msg_type == "PONG":
        print("PONG  (liveness reply)")
    elif msg.msg_type == "EXIT":
        print(
            f"EXIT  gateway is closing the session "
            f"(reason={msg.fields.get('REASON', '?')})",
            file=sys.stderr,
        )
    elif msg.msg_type == "ERR":
        # RALF's error field is DETAIL, not MSG (CALF's error field name) --
        # see protocol reference §10.
        code = msg.fields.get("CODE", "?")
        detail = msg.fields.get("DETAIL", "")
        print(f"ERR   {code}: {detail}", file=sys.stderr)
    else:
        print(f"{msg.msg_type} {msg.fields}")


def main() -> None:
    parser = argparse.ArgumentParser(
        description=(
            "RALF subscriber example: role-entitled channel subscription "
            "plus a de-duplicated per-symbol executed-volume tally."
        )
    )
    parser.add_argument("--host", default="127.0.0.1")
    parser.add_argument("--port", type=int, default=5580)
    parser.add_argument("--client", default="ext-client-01")
    parser.add_argument(
        "--role",
        default="CLEARING",
        choices=["CLEARING", "DROP_COPY", "AUDIT"],
    )
    parser.add_argument(
        "--channels",
        default=None,
        help="Comma-separated channels to subscribe; defaults to every "
        "channel --role is entitled to (CLEARING->CLEARING, "
        "DROP_COPY->DROP_COPY, AUDIT->all three). Requesting a channel "
        "outside the role's entitlement will get ERR|CODE=ENTITLEMENT_DENIED.",
    )
    parser.add_argument("--symbols", default="*", help="Symbol scope for SUB")
    parser.add_argument("--lastseq", default="0")
    args = parser.parse_args()

    channels = (
        [c.strip().upper() for c in args.channels.split(",") if c.strip()]
        if args.channels
        else list(_ROLE_CHANNELS[args.role])
    )

    seq_tracker = ChannelSequenceTracker()
    exec_tracker = ExecTracker()

    with socket.create_connection((args.host, args.port), timeout=5) as sock:
        reader = LineReader(sock)
        send_line(
            sock,
            "HELLO",
            {
                "CLIENT": args.client,
                "PROTO": "RALF1",
                "ROLE": args.role,
                "LASTSEQ": args.lastseq,
            },
        )

        welcome = parse_ralf_line(reader.recv_line())
        print(f"WELCOME: type={welcome.msg_type} fields={welcome.fields}")

        subscribe(sock, channels, args.symbols)
        print(f"Subscribed CH={channels} SYM={args.symbols} as role={args.role}")

        connection_error: Exception | None = None
        try:
            while True:
                msg = parse_ralf_line(reader.recv_line())
                _handle_message(msg, seq_tracker, exec_tracker)
        except KeyboardInterrupt:
            print("\ninterrupted, closing connection", file=sys.stderr)
        except (RuntimeError, OSError) as exc:
            connection_error = exc
            print(f"connection lost: {exc}", file=sys.stderr)

    print()
    print(exec_tracker.summary())
    if connection_error is not None:
        sys.exit(1)


if __name__ == "__main__":
    main()
